site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+metadata:
+  name: foo-grpcroute
+spec:
+  parentRefs:
+  - name: my-gateway
+  hostnames:
+  - foo.com
+  rules:
+    matches:
+    - method:
+        service: foo.bar
+    backendRefs:
+    - name: grpc-backend
+      port: 50051
+---
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: foo-httproute
+spec:
+  parentRefs:
+  - name: my-gateway
+  hostnames:
+  - foo.com
+  rules:
+    matches:
+    - path:
+        type: PathPrefix
+        value: /foo
+    backendRefs:
+    - name: http-backend
+      port: 80
+```
+
+According to the above scheme, we translate the gRPC method matcher to an HTTP
+path matcher of type `PathPrefix` and value `/foo.bar/`. Following the HTTPRoute
+precedence rules, `/foo.bar/` has more characters than `/foo`, so it has higher
+precedence. Hence, the GRPCRoute's rule must be evaluated before the HTTPRoute's
+rule.
+
+Accordingly, an xDS server should create something like the following
+[`RouteConfiguration`](https://github.com/envoyproxy/envoy/blob/379a7567b0d15cc9d0806052b82670c8cad76fc5/api/envoy/config/route/v3/route.proto#L26):
+
+```json
+{
+  "virtual_hosts": [
+    "domains": ["foo.com"],
+    "routes": [
+      {
+        "prefix": "/foo.bar/",
+        "route": {
+          "cluster": "grpc-backend-cluster"
+        }
+      },
+      {
+        "prefix": "/foo",
+        "route": {
+          "cluster": "http-backend-cluster"
+        }
+      }
+    ]
+  ]
+}
+```
+
+## API
+
+The API deviates from `HTTPRoute` where it results in a better UX for gRPC
+users, while mirroring it in all other cases.
+
+### Example `GRPCRoute`
+
+```yaml
+kind: GRPCRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: foo-grpcroute
+spec:
+  parentRefs:
+  - name: my-gateway
+  hostnames:
+  - foo.com
+  - bar.com
+  rules:
+  - matches:
+      method:
+        service: helloworld.Greeter
+        method:  SayHello
+      headers:
+      - type: Exact
+        name: magic
+        value: foo
+
+    filters:
+    - type: RequestHeaderModifierFilter
+      add:
+        - name: my-header
+          value: foo
+
+    - type: RequestMirrorPolicyFilter
+      destination:
+        backendRef:
+          name: mirror-svc
+
+    backendRefs:
+    - name: foo-v1
+      weight: 90
+    - name: foo-v2
+      weight: 10
+```
+
+### Structs
+
+```go
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=gateway-api
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Hostnames",type=string,JSONPath=`.spec.hostnames`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// GRPCRoute provides a way to route gRPC requests. This includes the capability
+// to match requests by hostname, service, method, or header. Filters can be
+// used to specify additional processing steps. Backends specify where matching
+// requests should be routed.
+type GRPCRoute struct {
```

Comment:
What is the conformance level for this route?

Reply:
Since we only have one conformance profile at the moment, essentially, I'd say its Extended.

Reply:
Added.